### PR TITLE
Improve and use HasEventHandlerSignature helper

### PR DESCRIPTION
**Pull request by Evangelink**
When reviewing the code to start the implementation of #488 I have realized we have a lot of methods doing the detection of event-handler signature but they were not all in sync so I have removed them all and factorized them in one helper.

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs [hunk 1]
```diff
@@ -95,39 +94,14 @@ public override void Initialize(AnalysisContext analysisContext)
...
-                    bool IsEventArgsParameter(IParameterSymbol parameter)
-                    {
-                        var type = parameter.Type;
-                        if (IsAssignableTo(type, eventArgs, context.Compilation))
-                        {
-                            return true;
-                        }
-
-                        // FxCop compat: Struct with name ending with "EventArgs" are allowed.
-                        if (type.IsValueType)
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs]: For UWP we not only have struct but also class so remove this check.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs [hunk 2]
```diff
@@ -95,39 +94,14 @@ public override void Initialize(AnalysisContext analysisContext)
...
-                        {
-                            return type.Name.EndsWith("EventArgs", StringComparison.Ordinal);
-                        }
-
-                        return false;
-                    }
-
-                    bool IsValidNonGenericEventHandler(IMethodSymbol delegateInvokeMethod)
-                    {
-                        return delegateInvokeMethod.ReturnsVoid &&
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs]: A couple of helpers were checking for the `void` return type, I don't know whether that's best to include this in the helper or not.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs [hunk 3]
```diff
@@ -95,39 +94,14 @@ public override void Initialize(AnalysisContext analysisContext)
                     static bool IsDelegateTypeWithInvokeMethod(INamedTypeSymbol namedType) =>
                         namedType.TypeKind == TypeKind.Delegate && namedType.DelegateInvokeMethod != null;
 
-                    bool IsEventArgsParameter(IParameterSymbol parameter)
-                    {
-                        var type = parameter.Type;
-                        if (IsAssignableTo(type, eventArgs, context.Compilation))
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs:101]: Can we delete this protected abstract method now?

Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs:101]: Why not?

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs:101]: I meant can you now delete `IsAssignableTo`?

Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.cs:101]: That's why I love a good review!

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.Fixer.cs [hunk 4]
```diff
@@ -1,15 +1,18 @@
...
 using System.Threading.Tasks;
 
 namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
 {
     /// <summary>
     /// CA1003: Use generic event handler instances
     /// </summary>
-    public abstract class UseGenericEventHandlerInstancesFixer : CodeFixProvider
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class UseGenericEventHandlerInstancesFixer : CodeFixProvider
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/UseGenericEventHandlerInstances.Fixer.cs:15]: We should probably just delete this file - it wouldn't be hard to add it back if/when we add a fixer.

**Comment by Evangelink** [comment]
@mavasani PR ready to be merged

**Pull request merged**